src/core/net/ip6.cpp

```diff
@@ -1068,6 +1072,26 @@ Error Ip6::SendRaw(OwnedPtr<Message> aMessagePtr)
 
     if (header.GetDestination().IsMulticast())
     {
+#if OPENTHREAD_CONFIG_FILTER_LOCAL_ORIGINATED_MULTICAST_PACKETS
+        // For multicast packet originating from untrusted source on host and passed to OT stack,
+        // only forward to thread network if destination is in the multicast listener table.
+        if (aMessagePtr->IsOriginHostUntrusted() && Get<ThreadNetif>().HasUnicastAddress(header.GetSource()))
+        {
+#if OPENTHREAD_CONFIG_BACKBONE_ROUTER_MULTICAST_ROUTING_ENABLE && OPENTHREAD_FTD && \
+    OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+            forwardThread = (header.GetDestination().IsMulticastLargerThanRealmLocal() &&
+                             Get<BackboneRouter::MulticastListenersTable>().HasListener(header.GetDestination()));
+#else
+            forwardThread = false;
+#endif
+            if (!forwardThread)
+            {
+                LogInfo("Dropping host originated multicast packet when destination is not MLR subscribed.");
+            }
+            VerifyOrExit(forwardThread, error = kErrorDrop);
+        }
+#endif
+
```

Comment:
With this feature we can decide not to forward this message over the thread mesh (to other nodes) which sounds resoanable. But the `VerifyOrExit()` check here, we also disallow it to be received by the device itself (since we skipp calling `HandleDatagram()`).

What if this device itself is subscribed to the destination of the multicast message? Do we need to receive/process this multicast? Or should we drop it here as the desired behavior? (I don't have any  answer/suggestion myself, just raising it so we consider and think about this situation)

Related to this, we have `otMessageSetMulticastLoopEnabled()` which allows caller to inidcate whether we want to allow multicast loops.

If the goal is to only disallow forwarding over thread mesh, we can consider moving this new code block to `HandleDatagram()` method instead?

Reply:
Letting `otMessageSetMulticastLoopEnabled()` decide if multicast loops are allowed for host originated packets sounds good to me, I'll update this PR to move the code to `HandleDatagram()`.